sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala

```diff
@@ -117,37 +311,49 @@ object DeduplicateRelations extends Rule[LogicalPlan] {
           }
         }
 
-        val planWithNewSubquery = plan.transformExpressions {
-          case subquery: SubqueryExpression =>
-            val (renewed, changed) = renewDuplicatedRelations(existingRelations, subquery.plan)
-            if (changed) planChanged = true
-            subquery.withNewPlan(renewed)
-        }
+        val (newPlan, changed) = getNewPlanWithNewChildren(existingRelations, newChildren.toArray,
+          plan, planChanged)
+        planChanged |= changed
+        newPlan
+      } else {
+        plan
+      }
+      (newPlan, planChanged)
+  }
 
-        if (planChanged) {
-          if (planWithNewSubquery.childrenResolved) {
-            val planWithNewChildren = planWithNewSubquery.withNewChildren(newChildren.toSeq)
-            val attrMap = AttributeMap(
-              plan
-                .children
-                .flatMap(_.output).zip(newChildren.flatMap(_.output))
-                .filter { case (a1, a2) => a1.exprId != a2.exprId }
-            )
-            if (attrMap.isEmpty) {
-              planWithNewChildren
-            } else {
-              planWithNewChildren.rewriteAttrs(attrMap)
-            }
-          } else {
-            planWithNewSubquery.withNewChildren(newChildren.toSeq)
-          }
-        } else {
+  private def getNewPlanWithNewChildren(
```

Comment:
can we just add a big function `def getNewPlanWithRelationDeduplicated`, which renew duplicated relations in the children and subqueries?

Reply:
why is it hard? The original code can handle LogicalPlan with any number of children, and it can definitely take care of plans that only have one child.